robyn/openapi.py

```diff
@@ -0,0 +1,271 @@
+    openapi_spec: dict = field(init=False)
+
+    def __post_init__(self):
+        """
+        Initializes the openapi_spec dict
+        """
+        self.openapi_spec = {
+            "openapi": "3.0.0",
+            "info": asdict(self.info),
+            "paths": {},
+            "components": asdict(self.info.components),
+            "servers": [asdict(server) for server in self.info.servers],
+            "externalDocs": asdict(self.info.externalDocs) if self.info.externalDocs.url else None,
+        }
+
+    def add_openapi_path_obj(self, route_type: str, endpoint: str, openapi_summary: str, openapi_tags: list, signature: Signature = None):
+        """
+        Adds the given path to openapi spec
+
+        :param route_type: the http method as string (get, post ...)
+        :param endpoint: the endpoint to be added
+        :param openapi_summary: short summary of the endpoint (to be fetched from the endpoint defenition by default)
+        :param openapi_tags: tags -- for grouping of endpoints
+        :param signature: the function signature -- to grab the typed dict annotations for query params
+        """
+
+        query_params = None
+
+        if signature and "query_params" in signature.parameters:
+            query_params = signature.parameters["query_params"].default
+
+        modified_endpoint, path_obj = self.get_path_obj(endpoint, openapi_summary, openapi_tags, query_params)
+
+        if modified_endpoint not in self.openapi_spec["paths"]:
+            self.openapi_spec["paths"][modified_endpoint] = {}
+        self.openapi_spec["paths"][modified_endpoint][route_type] = path_obj
+
+    def add_subrouter_paths(self, subrouter_openapi):
+        """
+        Adds the subrouter paths to main router's openapi specs
+
+        :param subrouter_openapi: the OpenAPI object of the current subrouter
+        """
+        paths = subrouter_openapi.openapi_spec["paths"]
+
+        for path in paths:
+            self.openapi_spec["paths"][path] = paths[path]
+
+    def get_path_obj(self, endpoint: str, summary: str, tags: list, query_params: TypedDict = None):
+        """
+        Get the "path" openapi object according to spec
+
+        :param endpoint: the endpoint to be added
+        :param summary: short summary of the endpoint (to be fetched from the endpoint defenition by default)
+        :param tags: tags -- for grouping of endpoints
+        :param query_params: query params for the function
+
+        :return: the "path" openapi object according to spec
+        """
+        # robyn has paths like /:url/:etc whereas openapi requires path like /{url}/{path}
+        # this function is used for converting path params to the required form
+        # initialized with endpoint for handling endpoints without path params
+        endpoint_with_path_params_wrapped_in_braces = endpoint
+
+        path_params = endpoint.split(":")
+        openapi_parameter_object = []
+
+        if len(path_params) > 1:
```

Comment:
I feel we can write it a bit better.

Reply:
I have updated the implementation, and used better variable names, please check. :)